src/viam/examples/modules/complex/gizmo/api.hpp

```diff
@@ -64,8 +64,8 @@ class GizmoClient : public Gizmo {
 // `GizmoServer` is the gRPC server implementation of a `Gizmo` component.
 class GizmoServer : public ResourceServer, public GizmoService::Service {
    public:
-    GizmoServer();
-    explicit GizmoServer(std::shared_ptr<ResourceManager> manager);
+    GizmoServer(std::shared_ptr<Server> server);
+    explicit GizmoServer(std::shared_ptr<ResourceManager> manager, std::shared_ptr<Server> server);
```

Comment:
Various thoughts here:

- Feels boilerplate-ish if every `[TypeOfResource]Server` needs to declare two nearly identical constructors. What's the use-case for each? Is there a way we could combine them?

```
GizmoServer(std::shared_ptr<Server> server, std::shared_ptr<ResourceManager> manager = {}) : 
    ResourceServer(server, manager ? manager : std::make_shared<ResourceManager>()) {}
```

That still feels chatty. Could `ResourceServer` do the same trick instead?

- If there are going to be two signatures, I'd put the common parameter first (`std::shared_ptr<Server>`).

- I see a lot of places where `ResourceManager` and `Server` travel around together. Is there a missing abstraction? Why is there only sometimes a `ResourceManager`?

Reply:
Yeah now that you mention it, I see no usages of the `[TypeOfResource]Server(std::shared_ptr<Server> server)` constructors. In other words, I think you are correct that we always create resource servers with both a `ResourceManager` and `Server`. I _could_ remove all the `[TypeOfResource]Server(std::shared_ptr<Server> server)` constructors, but that would be hypothetically backward breaking. In any case, I'll try an approach like what you've suggested with default values for `manager` and a ternary operator to make one if none is passed in (hopefully at the `ResourceServerBase` level to be as simple as possible).

Reply:
Actually I'll hold off on this until @stuqdog weighs in on whether we can remove `[TypeOfResource]Server` ctors that take no resource manager.

Reply:
> whether we can remove [TypeOfResource]Server ctors that take no resource manager.
I think this is fine. I agree we don't seem to be using the `manager`-less version anywhere, it's easy enough to make a `ResourceManager` if we need to, and the current structure is a bit noisy.